KOGITO-8414 Added knative-service-discovery.adoc

hbelmiro opened the pull request:
<!-- Please don't forget your JIRA link -->
**JIRA:** https://issues.redhat.com/browse/KOGITO-8414

This guide is for Knative custom function, which is implemented in https://github.com/kiegroup/kogito-runtimes/pull/2711

The example used by this guide is implemented in https://github.com/kiegroup/kogito-examples/pull/1526

**This PR should be merged only after the above PRs are merged.**

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributions doc](https://github.com/kiegroup/kogito-docs/blob/main/CONTRIBUTING.md)
- [x] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [x] Pull Request title contains the target branch if not targeting main: `[0.9.x] KOGITO-XYZ Subject`
- [x] The nav.adoc file has a link to this guide in the proper category
- [x] The index.adoc file has a card to this guide in the proper category, with a meaningful description

<details>
<summary>
How to backport a pull request to a different branch?
</summary>

In order to automatically create a **backporting pull request** please add one or more labels having the following format `backport-<branch-name>`, where `<branch-name>` is the name of the branch where the pull request must be backported to (e.g., `backport-7.67.x` to backport the original PR to the `7.67.x` branch).

> **NOTE**: **backporting** is an action aiming to move a change (usually a commit) from a branch (usually the main one) to another one, which is generally referring to a still maintained release branch. Keeping it simple: it is about to move a specific change or a set of them from one branch to another.

Once the original pull request is successfully merged, the automated action will create one backporting pull request per each label (with the previous format) that has been added.

If something goes wrong, the author will be notified and at this point a manual backporting is needed.

> **NOTE**: this automated backporting is triggered whenever a pull request on `main` branch is labeled or closed, but both conditions must be satisfied to get the new PR created.
</details>

spolti commented on serverlessworkflow/modules/ROOT/pages/cloud/knative-service-discovery.adoc:
```diff
@@ -0,0 +1,292 @@
+
+[source,xml]
+----
+<dependency>
```
please do not forget to update the existing SD doc.

hbelmiro replied:
Done.

kaldesai commented on serverlessworkflow/modules/ROOT/pages/cloud/knative-service-discovery.adoc:
```diff
@@ -0,0 +1,292 @@
+
+This document describes how to call Knative services using {context} custom functions. The procedure described in this document is based on the link:{kogito_sw_examples_url}/serverless-workflow-custom-function-knative[`serverless-workflow-custom-function-knative`] example application.
+
+The testing procedure in this document is based on the serverless-workflow-examples/serverless-workflow-custom-function-knative example application. You can access this example application in link:{kogito_sw_examples_url}/serverless-workflow-custom-function-knative[Kogito Examples] GitHub repository.
```
```suggestion
The testing procedure in this document is based on the serverless-workflow-examples/serverless-workflow-custom-function-knative example application. You can access this example application in the link:{kogito_sw_examples_url}/serverless-workflow-custom-function-knative[Kogito Examples] GitHub repository.
```

kaldesai commented on serverlessworkflow/modules/ROOT/pages/cloud/knative-service-discovery.adoc:
```diff
@@ -0,0 +1,292 @@
+[[about-namespaces]]
+=== About namespaces
+
+Note that in the above example you declared only the name of the service you wanted to call, but not a namespace. In this case, {product_name} will look for a Knative service in the same namespace the workflow service is running.
```
```suggestion
Note that in the above example, you declared only the name of the service you wanted to call, but not a namespace. In this case, {product_name} will look for a Knative service in the same namespace the workflow service is running.
```

kaldesai commented on serverlessworkflow/modules/ROOT/pages/cloud/knative-service-discovery.adoc:
```diff
@@ -0,0 +1,292 @@
+
+You can specify the HTTP resource path by adding the `path` property to the function's metadata.
+
+For instance, given the `/path/to/my/resource` HTTP resource path, you can define your function as:
```
```suggestion
For an instance, given the `/path/to/my/resource` HTTP resource path, you can define your function as follows:
```

hbelmiro replied:
"For **_an_** instance"? Are you sure?
My intention is to say the same as "For example".

kaldesai commented on serverlessworkflow/modules/ROOT/pages/cloud/knative-service-discovery.adoc:
```diff
@@ -0,0 +1,292 @@
+
+By default, the Knative service must respond within 10 seconds. You can use the `kogito.addon.knative-serving.request-timeout` property to configure this value.
+
+For instance, if you want to reduce the request timeout to 5 seconds, you must add the following to your `application.properties` file:
```
```suggestion
For an instance, if you want to reduce the request timeout to 5 seconds, you must add the following to your `application.properties` file:
```

hbelmiro replied:
Same as https://github.com/kiegroup/kogito-docs/pull/276#discussion_r1065875482

kaldesai reviewed (approved):
@hbelmiro I have very minor suggestions. Your content looks good to me. Very well written! Thank you!

hbelmiro commented:
@ricardozanini as we discussed offline, I've changed the structure to match camel-routes-integration.adoc and moved the guide to the "Integrations" menu.

radtriste commented on serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc:
```diff
@@ -0,0 +1,335 @@
+:description: Explain what is and how the Knative service discovery works
+:keywords: kogito, workflow, quarkus, serverless, service-discovery, knative
+
+This document describes how to call Knative services using {context} custom functions. The procedure described in this document is based on the link:{kogito_sw_examples_url}/serverless-workflow-custom-function-knative[`serverless-workflow-custom-function-knative`] example application.
+
+The testing procedure in this document is based on the serverless-workflow-examples/serverless-workflow-custom-function-knative example application. You can access this example application in the link:{kogito_sw_examples_url}/serverless-workflow-custom-function-knative[Kogito Examples] GitHub repository.
```
Seems like you are repeating yourself here :)

hbelmiro replied:
I removed the second one ;)

radtriste commented on serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc:
```diff
@@ -0,0 +1,335 @@
+The testing procedure in this document is based on the serverless-workflow-examples/serverless-workflow-custom-function-knative example application. You can access this example application in the link:{kogito_sw_examples_url}/serverless-workflow-custom-function-knative[Kogito Examples] GitHub repository.
+
+[[enabling-knative-custom-functions]]
+== Enabing Knative {context} custom functions
```
```suggestion
== Enabling Knative {context} custom functions
```

radtriste commented on serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc:
```diff
@@ -0,0 +1,335 @@
+
+.Prerequisites
+
+include::../cloud/common/_prerequisites.adoc[]
```
Missing attributes:
```
{environment_prereq}

kubectl {kubectl_prereq}
```

radtriste replied:
also do we really need all those prerequisites ?
In the procedure, you just mention to add the new dependency but you never show how to do it with Maven, quarkus CLI or kn CLI ...

hbelmiro replied:
Yeah, we don't need them.
Removed.

radtriste commented on serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc:
```diff
@@ -0,0 +1,335 @@
+
+include::../cloud/common/_prerequisites.adoc[]
+
+.Procedure
```
Display issue

hbelmiro replied:
Fixed.

radtriste reviewed (commented):
The whole first part explaining the custom functions should be in https://kiegroup.github.io/kogito-docs/serverlessworkflow/latest/core/custom-functions-support.html IMO
and only the "Example" should be in this page

MarianMacik commented on serverlessworkflow/modules/ROOT/pages/integrations/custom-functions-knative.adoc:
```diff
@@ -0,0 +1,335 @@
+
+.Procedure
+
+. Add the `knative-serving` add-on dependency to your link:{kogito_sw_examples_url}/serverless-workflow-custom-function-knative/workflow[workflow] project.
```
The example could be mentioned earlier before the Prerequisites section, maybe. Something like we have in other docs, e.g. these steps are based on the serverless-workflow-custom-function-knative example.

hbelmiro replied:
This is mentioned in the guide's first paragraph. Do you think it's necessary to mention it again here?

MarianMacik replied:
@hbelmiro OK, I probably overlooked it. Thanks!

hbelmiro commented:
> The whole first part explaining the custom functions should be in https://kiegroup.github.io/kogito-docs/serverlessworkflow/latest/core/custom-functions-support.html IMO and only the "Example" should be in this page

The first part is explaining **Knative** custom functions (a specific type of custom function), while https://kiegroup.github.io/kogito-docs/serverlessworkflow/latest/core/custom-functions-support.html is for custom functions in general. It explains how to create a custom function.

ricardozanini commented:
I agree with @radtriste, @hbelmiro. It would be nice to keep the ubiquity among the custom functions documentation.

hbelmiro commented:
> > > The whole first part explaining the custom functions should be in https://kiegroup.github.io/kogito-docs/serverlessworkflow/latest/core/custom-functions-support.html IMO and only the "Example" should be in this page
> > 
> > 
> > @radtriste The first part is explaining **Knative** custom functions (a specific type of custom function), while https://kiegroup.github.io/kogito-docs/serverlessworkflow/latest/core/custom-functions-support.html is for custom functions in general. It explains how to create a custom function.
> 
> well we are explaining there the Camel custom functions. could be good to centralize the custom functions type on that page. And then have a specific example in Integrations section, like @ricardozanini did for Camel

Done @radtriste.

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc:
```diff
@@ -297,10 +297,179 @@ In this last example, the `jq` expression result is used as the `body` argument
 === Function return values
 
-The Camel route is responsible to produce the return value in a way that the workflow can understand. The following are considered valid objects:
+The Camel route is responsible to produce the return value in a way that the workflow can understand.The following are considered valid objects:
```
```suggestion
The Camel route is responsible to produce the return value in a way that the workflow can understand. The following are considered valid objects:
```

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc:
```diff
@@ -297,10 +297,179 @@ In this last example, the `jq` expression result is used as the `body` argument
+[#knative-custom-function]
+== Knative custom function
+
+{product_name} provides an implementation of custom function through the `knative-serving` add-on to invoke Knative services.It allows you to have a static URI, defining a Knative service, which is used to perform HTTP requests.The Knative service defined in the URI is queried in the current Knative cluster and translated to a valid URL.
```
```suggestion
{product_name} provides an implementation of a custom function through the `knative-serving` add-on to invoke Knative services. It allows you to have a static URI, defining a Knative service, which is used to perform HTTP requests. The Knative service defined in the URI is queried in the current Knative cluster and translated to a valid URL.
```

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc:
```diff
@@ -297,10 +297,179 @@ In this last example, the `jq` expression result is used as the `body` argument
+
+[NOTE]
+====
+This section briefly exemplifies how to define and use Knative custom functions within your workflow application.For more information, see xref:integrations/custom-functions-knative.adoc[Invoking Knative services from Serverless Workflow custom functions].
```
```suggestion
This section briefly exemplifies how to define and use Knative custom functions within your workflow application. For more information, see xref:integrations/custom-functions-knative.adoc[Invoking Knative services from Serverless Workflow custom functions].
```

radtriste commented on serverlessworkflow/modules/ROOT/nav.adoc:
```diff
@@ -64,6 +64,7 @@
 ** xref:cloud/build-and-deploy-with-serverless-operator-on-kubernetes.adoc[Buiding and deploying a  {context} application on Kubernetes using the {product_name} Serverless Operator]
 * Integrations
 ** xref:integrations/camel-routes-integration.adoc[Integrating with Camel Routes]
+** xref:integrations/custom-functions-knative.adoc[Invoking Knative services from {context} custom functions]
```
I would say that you can remove `custom functions` from the title

hbelmiro replied:
Done.

radtriste replied:
please also update navbar and index :)

radtriste reviewed (commented):
I prefer much more this version :)

Just one minor comment about the page title. I think it can be reduced.

MarianMacik reviewed (approved):
Looks good once navbar and index are updated.

hbelmiro commented:
> Looks good once navbar and index are updated.

Done.

radtriste commented on serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc:
```diff
@@ -301,6 +301,175 @@ The Camel route is responsible to produce the return value in a way that the wor
+
+[NOTE]
+====
+This section briefly exemplifies how to define and use Knative custom functions within your workflow application. For more information, see xref:integrations/custom-functions-knative.adoc[Invoking Knative services from Serverless Workflow custom functions].
```
```suggestion
This section briefly exemplifies how to define and use Knative custom functions within your workflow application. For more information, see xref:integrations/custom-functions-knative.adoc[Invoking Knative services from Serverless Workflow].
```
